src/v/cluster/rm_stm.cc

```diff
@@ -341,22 +384,39 @@ ss::future<tx_errc> rm_stm::commit_tx(
         _mem_state = mem_state{.term = _insync_term};
     }
 
-    if (_mem_state.preparing.contains(pid)) {
-        // it looks like a violation since tm_stm commits only after
-        // prepare succeeds but it's a legit rare situation:
-        //   * tm_stm failed during prepare
-        //   * its log update didn't pass
-        //   * on recovery it recommits its previous tx
-        //   * this commit (we're here) collides with "next" prepare
-        vlog(
-          clusterlog.error,
-          "Can't commit pid:{} - concurrent prepare operation on the same "
-          "session",
-          pid);
-        co_return tx_errc::conflict;
+    auto preparing_it = _mem_state.preparing.find(pid);
+    auto prepare_it = _log_state.prepared.find(pid);
+
+    if (preparing_it != _mem_state.preparing.end()) {
+        if (preparing_it->second > tx_seq) {
+            // - tm_stm & rm_stm failed during prepare
+            // - during recovery tm_stm recommits its previous tx
+            // - that commit (we're here) collides with "next" failed prepare
+            // it may happen only if the commit passed => acking
+            co_return tx_errc::none;
+        }
+
+        std::string msg;
```

Comment:
why not ss::sstring and ssx::format?

Reply:
I wasn't aware of this extension, fixed, thanks!